File: android/app/src/main/java/com/alexivanov/snapsell/ui/itemdetail/ItemDetailScreen.kt
package com.alexivanov.snapsell.ui.itemdetail

import android.content.Intent
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.net.toUri
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import coil3.compose.AsyncImage
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.R
import com.alexivanov.snapsell.data.remote.dto.Comp
import com.alexivanov.snapsell.data.remote.dto.SoldEstimateSource
import com.alexivanov.snapsell.domain.Money
import com.alexivanov.snapsell.ui.common.CutoutTile
import com.alexivanov.snapsell.ui.common.ErrorText
import com.alexivanov.snapsell.ui.common.LoadingBox
import com.alexivanov.snapsell.ui.common.LucideIcon
import com.alexivanov.snapsell.ui.common.MicroLabel
import com.alexivanov.snapsell.ui.common.OutlinedTag
import com.alexivanov.snapsell.ui.common.PrimaryButton
import com.alexivanov.snapsell.ui.common.Rule
import com.alexivanov.snapsell.ui.common.SecondaryButton
import com.alexivanov.snapsell.ui.common.SnapBottomBar
import com.alexivanov.snapsell.ui.common.SnapTextField
import com.alexivanov.snapsell.ui.common.SnapTopBar
import com.alexivanov.snapsell.ui.common.Spinner
import com.alexivanov.snapsell.ui.common.VRule
import com.alexivanov.snapsell.ui.theme.SnapType
import com.alexivanov.snapsell.ui.theme.snapColors
import java.io.File

@Composable
fun ItemDetailScreen(
    container: AppContainer,
    itemId: String,
    onHandOff: (String) -> Unit,
    onAddToBundle: (String) -> Unit,
    onBack: () -> Unit,
) {
    val vm: ItemDetailViewModel = viewModel(key = itemId) { ItemDetailViewModel(container, itemId) }
    val state by vm.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val c = snapColors

    Scaffold(
        containerColor = c.bg,
        topBar = { SnapTopBar(title = "Item", onBack = onBack) },
        bottomBar = {
            val item = state.item
            if (item != null) {
                SnapBottomBar {
                    SecondaryButton("Add to bundle", onClick = { onAddToBundle(item.id) })
                    PrimaryButton("List on Marketplace", onClick = { vm.createListing(onHandOff) }, modifier = Modifier.weight(1f))
                }
            }
        },
    ) { padding ->
        val item = state.item
        if (state.loading) {
            LoadingBox(Modifier.padding(padding))
            return@Scaffold
        }
        if (item == null) {
            Column(Modifier.padding(padding).padding(24.dp)) { Text("This item no longer exists.", style = SnapType.body, color = c.text) }
            return@Scaffold
        }
        val quote = item.quote

        Column(Modifier.padding(padding).fillMaxSize().verticalScroll(rememberScrollState())) {
            // 190dp white cutout band with a 2dp bottom divider.
            Box(Modifier.fillMaxWidth().height(190.dp).background(c.cutoutWhite), contentAlignment = Alignment.Center) {
                AsyncImage(model = File(item.photoPath), contentDescription = null, contentScale = ContentScale.Fit, modifier = Modifier.fillMaxSize().padding(12.dp))
            }
            Rule(thickness = 2.dp)

            Column(Modifier.padding(horizontal = 16.dp)) {
                Text(item.name.ifBlank { "Unidentified item" }, style = SnapType.sectionHead, color = c.text, modifier = Modifier.padding(top = 16.dp))
                Text(
                    listOfNotNull(item.brand, item.model, item.condition.label).joinToString(" · "),
                    style = SnapType.fieldLabel,
                    color = c.text.copy(alpha = 0.6f),
                    modifier = Modifier.padding(top = 4.dp),
                )

                // Price row, baseline-aligned: SUGGESTED over the 46sp figure, Final price field beside it.
                Row(Modifier.fillMaxWidth().padding(top = 20.dp), verticalAlignment = Alignment.Bottom) {
                    Column(Modifier.weight(1f)) {
                        MicroLabel("Suggested")
                        val suggested = quote?.suggestedPrice
                        Text(
                            when {
                                suggested != null -> Money.usd(suggested)
                                quote != null -> "No comps"
                                else -> "—"
                            },
                            style = SnapType.displayPrice,
                            color = c.text,
                        )
                    }
                    SnapTextField(
                        value = state.priceText,
                        onValueChange = vm::editPrice,
                        label = "Final price",
                        placeholder = quote?.suggestedPrice?.let(Money::plain) ?: "0",
                        prefix = "$",
                        minHeight = 48.dp,
                        textStyle = SnapType.fieldValueBold,
                        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
                        modifier = Modifier.width(126.dp),
                    )
                }
                ErrorText(state.error, Modifier.padding(top = 8.dp))
                if (quote == null) {
                    Row(Modifier.padding(top = 12.dp), verticalAlignment = Alignment.CenterVertically) {
                        SecondaryButton("Get a price", onClick = vm::reprice, enabled = !state.pricing, minHeight = 44.dp)
                        if (state.pricing) { Spacer(Modifier.width(12.dp)); Spinner() }
                    }
                }
            }

            if (quote != null) {
                // Stat strip: 2dp top rule, 1dp bottom rule, cells divided by 1dp verticals.
                Column(Modifier.padding(top = 20.dp)) {
                    Rule(thickness = 2.dp)
                    Row(Modifier.fillMaxWidth().height(IntrinsicSize.Min)) {
                        StatCell("Asking median", quote.askingMedian, Modifier.weight(1f))
                        VRule()
                        StatCell("Low", quote.askingLow, Modifier.weight(1f))
                        VRule()
                        StatCell("High", quote.askingHigh, Modifier.weight(1f))
                    }
                    Rule(thickness = 1.dp)
                }

                Column(Modifier.padding(horizontal = 16.dp)) {
                    val cond = item.condition.label.lowercase()
                    Text(
                        "From ${quote.compCount} active ${if (quote.conditionFiltered) "$cond " else ""}listings · local factor ${(quote.localSaleFactor * 100).toInt()}%",
                        style = SnapType.fieldLabel.copy(fontSize = 11.5.sp),
                        color = c.text.copy(alpha = 0.6f),
                        modifier = Modifier.padding(top = 10.dp),
                    )
                    if (!quote.conditionFiltered) {
                        Row(Modifier.fillMaxWidth().padding(top = 12.dp).background(c.surface)) {
                            Box(Modifier.width(3.dp).height(IntrinsicSize.Min).fillMaxSize().background(c.accent))
                            Text(
                                "No $cond-condition comps were found, so these ${quote.compCount} use every condition. Treat the number as rough.",
                                style = SnapType.bodySmall,
                                color = c.text,
                                modifier = Modifier.padding(horizontal = 12.dp, vertical = 10.dp),
                            )
                        }
                    }

                    if (quote.comps.isNotEmpty()) {
                        MicroLabel("Comparable listings", Modifier.padding(top = 24.dp, bottom = 8.dp))
                        Rule(thickness = 2.dp)
                        quote.comps.forEach { comp ->
                            CompRow(comp) { runCatching { context.startActivity(Intent(Intent.ACTION_VIEW, comp.url.toUri())) } }
                            Rule(thickness = 1.dp)
                        }
                    }

                    quote.estimatedSold?.let { est ->
                        Column(Modifier.fillMaxWidth().padding(top = 24.dp).border(2.dp, c.text).padding(16.dp)) {
                            MicroLabel("Estimated sold range")
                            Text("${Money.usd(est.low)}–${Money.usd(est.high)}", style = SnapType.soldRange, color = c.text, modifier = Modifier.padding(top = 4.dp))
                            val isLlm = est.source == SoldEstimateSource.LLM_ESTIMATE
                            OutlinedTag(
                                if (isLlm) "Estimate, not sales data" else "Source: ${est.source.name.lowercase().replace('_', ' ')}",
                                Modifier.padding(top = 10.dp),
                            )
                            Text(
                                if (isLlm) {
                                    "Modelled down from what sellers are asking. eBay does not publish completed-sale prices through the API."
                                } else {
                                    est.rationale ?: "Reported by ${est.source.name.lowercase().replace('_', ' ')}."
                                },
                                style = SnapType.bodySmall,
                                color = c.text.copy(alpha = 0.8f),
                                modifier = Modifier.padding(top = 10.dp),
                            )
                        }
                    }
                }
            }
            Spacer(Modifier.height(24.dp))
        }
    }
}

@Composable
private fun StatCell(label: String, value: Double?, modifier: Modifier) {
    val c = snapColors
    Column(modifier.padding(horizontal = 16.dp, vertical = 12.dp)) {
        MicroLabel(label, alpha = 0.55f, style = SnapType.statLabel)
        Text(value?.let(Money::usd) ?: "—", style = SnapType.statValue, color = c.text, modifier = Modifier.padding(top = 4.dp))
    }
}

/** 46dp thumbnail · two-line title + condition · TOTAL on the right · external-link icon. */
@Composable
private fun CompRow(comp: Comp, onClick: () -> Unit) {
    val c = snapColors
    Row(Modifier.fillMaxWidth().clickable(onClick = onClick).padding(vertical = 10.dp), verticalAlignment = Alignment.CenterVertically) {
        if (comp.imageUrl != null) {
            CutoutTile(comp.imageUrl, 46.dp, contentScale = ContentScale.Crop)
        } else {
            Box(Modifier.size(46.dp).background(c.neutral300))
        }
        Column(Modifier.weight(1f).padding(horizontal = 12.dp)) {
            Text(comp.title, style = SnapType.bodySmall.copy(fontSize = 12.5.sp, lineHeight = 16.sp), color = c.text, maxLines = 2, overflow = TextOverflow.Ellipsis)
            comp.condition?.let { Text(it, style = SnapType.fieldLabel.copy(fontSize = 11.sp), color = c.text.copy(alpha = 0.55f), modifier = Modifier.padding(top = 2.dp)) }
        }
        Column(horizontalAlignment = Alignment.End) {
            Text(Money.usd(comp.total), style = SnapType.rowTitle, color = c.text, textAlign = TextAlign.End)
            MicroLabel("Total")
        }
        LucideIcon(R.drawable.ic_lucide_external_link, "Open listing", Modifier.padding(start = 12.dp), size = 16.dp, tint = c.text.copy(alpha = 0.7f))
    }
}

File: ios/SnapSell/UI/Screens/ItemDetail/ItemDetailScreen.swift
import SafariServices
import SwiftUI

struct ItemDetailScreen: View {
    let itemId: String
    @Environment(AppContainer.self) private var app
    @Environment(Router.self) private var router
    @Environment(\.snap) private var c
    @State private var priceText = ""
    @State private var priceTouched = false
    @State private var pricing = false
    @State private var error: String?
    @State private var compURL: URL?

    private var item: Item? { app.inventory.item(itemId) }

    var body: some View {
        VStack(spacing: 0) {
            SnapTopBar("Item", onBack: { router.pop() })
            if let item {
                content(item)
                VStack(spacing: 0) {
                    Rule()
                    HStack(spacing: 8) {
                        SnapButton("Add to bundle", kind: .secondary, fullWidth: false) { router.push(.bundle(itemId: item.id)) }
                        SnapButton("List on Marketplace") { createListing(item) }
                    }
                    .padding(16)
                }
                .background(c.bg)
            } else {
                Text("This item no longer exists.").snap(SnapType.body).foregroundStyle(c.text).padding(24)
                Spacer()
            }
        }
        .onAppear { if !priceTouched { priceText = item?.finalPrice.map(Money.plain) ?? "" } }
        .onChange(of: item?.finalPrice) { _, v in if !priceTouched { priceText = v.map(Money.plain) ?? "" } }
        .sheet(item: $compURL) { url in SafariView(url: url).ignoresSafeArea() }
        .snapScreen()
    }

    private func content(_ item: Item) -> some View {
        let quote = item.quote
        return ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                ZStack { Color.white; StoredImage(file: item.photoFile).padding(12) }.frame(height: 190)
                Rule()
                VStack(alignment: .leading, spacing: 0) {
                    Text(item.name.isBlank ? "Unidentified item" : item.name).snap(SnapType.sectionHead).foregroundStyle(c.text).padding(.top, 16)
                    Text([item.brand, item.model, item.condition.label].compactMap { $0 }.joined(separator: " · "))
                        .snap(SnapType.fieldLabel).foregroundStyle(c.text.opacity(0.6))
                    HStack(alignment: .bottom, spacing: 16) {
                        VStack(alignment: .leading, spacing: 2) {
                            MicroLabel("Suggested")
                            Text(quote?.suggestedPrice.map(Money.usd) ?? (quote != nil ? "No comps" : "—")).snap(SnapType.displayPrice).foregroundStyle(c.text)
                        }
                        Spacer(minLength: 0)
                        VStack(alignment: .leading, spacing: 6) {
                            FieldLabel("Final price")
                            HStack(spacing: 4) {
                                Text("$").snap(SnapType.fieldValueBold).foregroundStyle(c.text)
                                TextField(quote?.suggestedPrice.map(Money.plain) ?? "0", text: $priceText)
                                    .keyboardType(.decimalPad).snapText(SnapType.fieldValueBold).foregroundStyle(c.text)
                                    .onChange(of: priceText) { _, t in editPrice(t) }
                            }
                            .padding(.horizontal, 12).frame(width: 140, height: 48)
                            .background(c.surface).overlay(Rectangle().stroke(c.text, lineWidth: SnapMetrics.rule))
                        }
                    }
                    .padding(.top, 20)
                    if let error { ErrorText(error).padding(.top, 8) }
                    if quote == nil {
                        HStack(spacing: 12) {
                            SnapButton("Get a price", kind: .secondary, enabled: !pricing, fullWidth: false) { reprice(item) }
                            if pricing { Spinner() }
                        }
                        .padding(.top, 12)
                    }
                }
                .padding(.horizontal, 16)
                if let quote {
                    VStack(spacing: 0) {
                        Rule()
                        HStack(spacing: 0) {
                            stat("Asking median", quote.askingMedian); VRule()
                            stat("Low", quote.askingLow); VRule()
                            stat("High", quote.askingHigh)
                        }
                        .fixedSize(horizontal: false, vertical: true)
                        Rule(color: c.divider).frame(height: 1)
                    }
                    .padding(.top, 20)
                    VStack(alignment: .leading, spacing: 0) {
                        let cond = item.condition.label.lowercased()
                        Text("From \(quote.compCount) active \(quote.conditionFiltered ? "\(cond) " : "")listings · local factor \(Int(quote.localSaleFactor * 100))%")
                            .font(.custom("Archivo-SemiBold", size: 11.5)).foregroundStyle(c.text.opacity(0.6)).padding(.top, 12)
                        if !quote.conditionFiltered {
                            HStack(spacing: 0) {
                                Rectangle().fill(c.accent).frame(width: 3)
                                Text("No \(cond)-condition comps were found, so these \(quote.compCount) use every condition. Treat the number as rough.")
                                    .snap(SnapType.bodySmall).foregroundStyle(c.text).padding(12)
                            }
                            .fixedSize(horizontal: false, vertical: true)
                            .background(c.surface).padding(.top, 12)
                        }
                        if !quote.comps.isEmpty {
                            MicroLabel("Comparable listings").padding(.top, 24).padding(.bottom, 8)
                            Rule()
                            ForEach(quote.comps) { comp in
                                CompRow(comp: comp) { compURL = URL(string: comp.url) }
                                Rule(color: c.divider).frame(height: 1)
                            }
                        }
                        if let est = quote.estimatedSold {
                            let isLlm = est.source == .llmEstimate
                            VStack(alignment: .leading, spacing: 0) {
                                MicroLabel("Estimated sold range")
                                Text("\(Money.usd(est.low))–\(Money.usd(est.high))").snap(SnapType.soldRange).foregroundStyle(c.text).padding(.top, 4)
                                OutlinedTag(isLlm ? "Estimate, not sales data" : "Source: \(est.source.humanName)").padding(.top, 10)
                                Text(isLlm ? "Modelled down from what sellers are asking. eBay does not publish completed-sale prices through the API."
                                     : (est.rationale ?? "Reported by \(est.source.humanName)."))
                                    .snap(SnapType.bodySmall).foregroundStyle(c.text.opacity(0.8)).padding(.top, 8)
                            }
                            .padding(16).frame(maxWidth: .infinity, alignment: .leading)
                            .overlay(Rectangle().stroke(c.text, lineWidth: SnapMetrics.rule))
                            .padding(.top, 24)
                        }
                    }
                    .padding(.horizontal, 16)
                }
                Spacer(minLength: 24)
            }
        }
        .scrollDismissesKeyboard(.interactively)
    }

    private func stat(_ label: String, _ value: Double?) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(label.uppercased()).snap(SnapType.statLabel).foregroundStyle(c.text.opacity(0.55))
            Text(value.map(Money.usd) ?? "—").snap(SnapType.statValue).foregroundStyle(c.text)
        }
        .padding(.horizontal, 16).padding(.vertical, 12)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    /// Typing a final price overrides the suggestion everywhere; clearing the field removes the override.
    private func editPrice(_ text: String) {
        priceTouched = true
        error = nil
        if text.isBlank {
            try? app.inventory.setFinalPrice(itemId: itemId, price: nil)
        } else if let p = Double(text), p >= 0 {
            try? app.inventory.setFinalPrice(itemId: itemId, price: p)
        }
    }

    private func reprice(_ item: Item) {
        pricing = true
        error = nil
        Task {
            switch await app.pricing.price(item: item.toDto(), localSaleFactor: app.settings.localSaleFactor) {
            case .success(let q): try? app.inventory.saveQuote(itemId: itemId, quote: q)
            case .failure(let e): error = e.message
            }
            pricing = false
        }
    }

    private func createListing(_ item: Item) {
        let price = Double(priceText) ?? item.finalPrice ?? item.quote?.suggestedPrice
        guard let price, price >= 0 else { error = "Set a price first."; return }
        do {
            let id = try app.inventory.createListing(
                kind: .single, itemIds: [itemId],
                title: item.listingTitle.nonBlank ?? item.name,
                description: item.listingDescription.nonBlank ?? defaultDescription(item),
                price: price
            )
            router.push(.handoff(listingId: id))
        } catch {
            self.error = AppError.describe(error).message
        }
    }

    private func defaultDescription(_ item: Item) -> String {
        var s = item.name
        if !item.attributes.isEmpty { s += ", " + item.attributes.joined(separator: ", ") }
        s += ". Condition: \(item.condition.label.lowercased())."
        if let notes = item.notes.nonBlank { s += " " + notes }
        s += " Local pickup."
        return s
    }
}

private struct CompRow: View {
    let comp: Comp
    let onTap: () -> Void
    @Environment(\.snap) private var c
    var body: some View {
        Button(action: onTap) {
            HStack(spacing: 12) {
                if let u = comp.imageUrl, let url = URL(string: u) {
                    AsyncImage(url: url) { img in img.resizable().scaledToFill() } placeholder: { c.neutral300 }
                        .frame(width: 46, height: 46).clipped().overlay(Rectangle().stroke(c.text, lineWidth: SnapMetrics.rule))
                } else {
                    c.neutral300.frame(width: 46, height: 46)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(comp.title).font(.custom("Archivo-Regular", size: 12.5)).foregroundStyle(c.text).lineLimit(2).multilineTextAlignment(.leading)
                    if let cond = comp.condition { Text(cond).font(.custom("Archivo-SemiBold", size: 11)).foregroundStyle(c.text.opacity(0.55)) }
                }
                Spacer(minLength: 8)
                VStack(alignment: .trailing, spacing: 2) {
                    Text(Money.usd(comp.total)).snap(SnapType.rowTitle).foregroundStyle(c.text)
                    MicroLabel("Total")
                }
                Image(systemName: "arrow.up.right.square").font(.system(size: 15)).foregroundStyle(c.text.opacity(0.7))
            }
            .padding(.vertical, 10)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

struct SafariView: UIViewControllerRepresentable {
    let url: URL
    func makeUIViewController(context: Context) -> SFSafariViewController { SFSafariViewController(url: url) }
    func updateUIViewController(_ vc: SFSafariViewController, context: Context) {}
}

extension URL: @retroactive Identifiable {
    public var id: String { absoluteString }
}
